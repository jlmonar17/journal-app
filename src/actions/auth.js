import Swal from "sweetalert2";
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";

export const smartLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());

        firebase
            .auth()
            .signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => {
                Swal.fire("Error", e.message, "error");
            })
            .finally(() => {
                dispatch(finishLoading());
            });
    };
};

// Asynchronus action, after resolve, it will dispatch another SYNCHRONUS action.
// Asynchronus actions return a callback. in difference to SYNCHRONUS actions.
export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase
            .auth()
            .signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => console.log(e));
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase
            .auth()
            .createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name });

                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => {
                Swal.fire("Error", e.message, "error");
            });
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());
    };
};

export const logout = () => ({ type: types.logout });
