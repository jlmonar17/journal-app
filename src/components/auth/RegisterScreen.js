import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { useDispatch, useSelector } from "react-redux";
import { removeError, setError } from "../../actions/ui";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";

export const RegisterScreen = () => {
    const dispatch = useDispatch();

    const { msgError } = useSelector((state) => state.ui);

    const { formValues, handleInputChange } = useForm({
        name: "JosÃ©",
        email: "jose@mail.com",
        password: "123456",
        password2: "123456",
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError("Name is required"));

            return false;
        }

        if (!validator.isEmail(email.trim())) {
            dispatch(setError("Email is not valid"));

            return false;
        }

        if (!validator.equals(password, password2) || password.length < 6) {
            dispatch(setError("Invalid password"));

            return false;
        }

        dispatch(removeError());
        return true;
    };

    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form onSubmit={handleRegister}>
                {msgError && (
                    <div className="auth__alert-error">{msgError}</div>
                )}

                <input
                    className="auth__input"
                    type="text"
                    placeholder="Name"
                    name="name"
                    autoComplete="off"
                    value={name}
                    onChange={handleInputChange}
                />
                <input
                    className="auth__input"
                    type="text"
                    placeholder="Email"
                    name="email"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    className="auth__input"
                    type="password"
                    placeholder="Password"
                    name="password"
                    autoComplete="off"
                    value={password}
                    onChange={handleInputChange}
                />
                <input
                    className="auth__input"
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    autoComplete="off"
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Login
                </button>

                <Link to="/auth/login" className="link">
                    Alredy registered?
                </Link>
            </form>
        </>
    );
};
