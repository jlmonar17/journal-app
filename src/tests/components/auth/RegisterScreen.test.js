import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { MemoryRouter } from "react-router-dom";
import { types } from "../../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

// We set initial state desired for Redux state.
const initialState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null,
    },
};

let store = mockStore(initialState);

describe("Tests for <RegisterScreen />", () => {
    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );

    test("Should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Should dispatch setError when trying to register with empty email", () => {
        const emailField = wrapper.find("input[name='email']");

        // Set empty string to email field
        emailField.simulate("change", {
            target: {
                value: "",
                name: "email",
            },
        });

        // Simulate form submit
        wrapper.find("form").simulate("submit", { preventDefault() {} });

        // Get all dispatched actions
        const actions = store.getActions();

        // We expect setError action was dispatched
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: "Email is not valid",
        });
    });

    test("Should show div error when there is an error during register", () => {
        const initialState = {
            auth: {},
            ui: {
                loading: false,
                // We specified an error message to simulate div with error in the wrapper
                msgError: "There is an error!",
            },
        };

        const store = mockStore(initialState);

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find(".auth__alert-error").exists()).toBe(true);
        expect(wrapper.find(".auth__alert-error").text().trim()).toBe(
            initialState.ui.msgError
        );
    });
});
