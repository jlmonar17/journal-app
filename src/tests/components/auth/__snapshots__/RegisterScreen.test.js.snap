// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <RegisterScreen /> Should render correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="JosÃ©"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="jose@mail.com"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Login
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Alredy registered?
    </a>
  </form>,
]
`;
